### PR TITLE
Quotas plugin

### PR DESCRIPTION
**Pull request by Natim**
Refs #173 
Based on #751 

- [x] Add tests.
- [x] Add a changelog entry.

r? @leplatrem 

- [x] Add a listener that keep track of storage changes in the bucket.
- [x] Add a view that display some stats about the bucket (collection_count, record_count and storage_size)
- [x] Add a listener that uses the configuration and the resulting storage_size to raise a Quota error if needed.

### CODE REVIEW

**Comment by tarekziade** [comment]
Like I said in https://github.com/Kinto/kinto/issues/173#issuecomment-236604658 I am still -1 on having a specific view for holding this information, which is of the same nature than last_modified.

**Comment by Natim** [comment]
That's not the goal of this PR, we can change that before merging if need be.

Right now, the solution proposed by @leplatrem in https://github.com/Kinto/kinto/issues/173#issuecomment-222102623 which seems to me to be cleaner.

Updating the bucket on each change in it might have a big impact on the kinto protocol and create a lot of HTTP 412. We had a similar discussion before with kinto-attachment and kinto-history and it seems to be safer not to update user data with plugins.

**Comment by tarekziade** [comment]
> Right now, I am trying to implement the solution proposed by @leplatrem in #173 (comment) which seems to me to be cleaner.

Please don't implement any code if we disagree on the design. This is a potential waste of time.

> Updating the bucket on each change in it might have a big impact on the kinto protocol and create a lot of HTTP 412

I am not sure why you want to update the bucket on each change. 

Please let's pause for a second here until we know exactly what we want. thx

**Comment by Natim** [comment]
> I am not sure why you want to update the bucket on each change.

I will update this PR with the behavior your are mentioning to show you :)

File: kinto/plugins/quotas/listener.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    # Update the bucket quotas values for each impacted record.
+    for (uri, obj_id, old, new) in targets:
+        if is_bucket(uri):
+            old = strip_stats_keys(old)
+            new = strip_stats_keys(new)
+
+        if action == 'create':
+            bucket_info['storage_size'] += record_size(new)
+            if resource_name == 'collection':
+                bucket_info['collection_count'] += 1
```
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: This code is really likely to be subject to race conditions. (If two people update records in the same bucket for instance.) @tarekziade was suggesting doing something like Redis CAS or we could maybe do it directly with a PostgreSQL request so that it become a consistent change.

File: kinto/core/resource/__init__.py [hunk 2]
```diff
@@ -342,10 +342,11 @@ def collection_delete(self):
         self._raise_412_if_modified()
 
         filters = self._extract_filters()
+        records, _ = self.model.get_records(filters=filters)
         deleted = self.model.delete_records(filters=filters)
```
Comment (glasserc) [inline kinto/core/resource/__init__.py]: Is it possible for there to be records returned that are then not removed by the delete (i.e. a race condition with another client)? Or is this all happening in a transaction?

Comment (Natim) [inline kinto/core/resource/__init__.py]: It is happening in a transaction thanks to the postgresql backend :)

File: kinto/plugins/quotas/listener.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+def on_resource_changed(event):
+    """
+    Everytime an object is created/changed/deleted, we update the
+    bucket counters. The entries are served as read-only in the
+    :mod:`kinto.plugins.quotas.views` module.
+
+    If a new object override the quotas, we reject the request.
+    """
+    payload = copy.deepcopy(event.payload)
```
Comment (glasserc) [inline kinto/plugins/quotas/listener.py]: Why do you copy the payload?

File: kinto/plugins/quotas/listener.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+    storage = event.request.registry.storage
+
+    if action == 'delete' and resource_name == 'bucket':
+        try:
+            storage.delete(parent_id=bucket_uri,
+                           collection_id='quota',
+                           object_id='bucket_info')
+        except RecordNotFoundError:  # Pragma: no cover
+            pass
```
Comment (glasserc) [inline kinto/plugins/quotas/listener.py]: Is the "no cover" pragma here because it's difficult to test, or because this is a "can't happen" scenario?

File: kinto/plugins/quotas/listener.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    # Update the bucket quotas values for each impacted record.
+    for (uri, obj_id, old, new) in targets:
+        old_size = record_size(old)
+        new_size = record_size(new)
+
+        if max_bytes_per_item is not None:
+            if new_size > max_bytes_per_item:
+                raise http_error(HTTPInsufficientStorage(),
+                                 errno=ERRORS.FORBIDDEN.value,
+                                 message=HTTPInsufficientStorage.explanation)
```
Comment (glasserc) [inline kinto/plugins/quotas/listener.py]: I think this can trigger even on a delete, which seems wrong to me.

File: kinto/plugins/quotas/listener.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                             message=HTTPInsufficientStorage.explanation)
+
+    if bucket_max_items is not None:
+        if bucket_info['record_count'] > bucket_max_items:
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=HTTPInsufficientStorage.explanation)
+
+    if collection_max_bytes is not None:
+        if collection_info['storage_size'] > collection_max_bytes:
```
Comment (glasserc) [inline kinto/plugins/quotas/listener.py]: Same concern about these checks being triggered on deletes -- I think we should always permit deletes, even if you're over quota.

File: kinto/plugins/quotas/listener.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=HTTPInsufficientStorage.explanation)
+
+    if collection_max_items is not None:
+        if collection_info['record_count'] > collection_max_items:
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=HTTPInsufficientStorage.explanation)
+
+    storage.update(parent_id=bucket_uri,
```
Comment (glasserc) [inline kinto/plugins/quotas/listener.py]: Could you tell me a little bit more about what `parent_id` means? I looked at the StorageBase class and I'm still kind of confused. In normal use, I would have assumed that it's a bucket ID, but here you're using a URL, and below you're using a collection URL. Where/how is this info getting stored? Is it in a special inaccessible Postgres table? Or is it in the same table user records are stored? If so, how do we protect the user from updating these records?

File: kinto/plugins/quotas/listener.py (hunk 3 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: You are right, it is a left over from the history plugin I took the idea from. No need for that here.

File: kinto/plugins/quotas/listener.py (hunk 4 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: I see it more as can't happen scenario but I can try to add a test for it.

File: kinto/plugins/quotas/listener.py (hunk 5 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: Yes, let me add a test for it then. :+1:

File: kinto/plugins/quotas/listener.py (hunk 7 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: You are correct that in case of the bucket we could use the bucket_id but for the collection, because we can have the same collection_id in two distinct buckets we have to use both the bucket_id and the collection id, but then you could also have the same id for a group and a collection the bucket, reason why in kinto (and not kinto,core) we made this convention of using the uri as the record unique id.

Comment (Natim) [inline kinto/plugins/quotas/listener.py]: > how do we protect the user from updating these records?

We prevent it because we don't have any API for the user to access it.
If we were to add this API we would need to make sure the user cannot modify it.

File: kinto/plugins/quotas/listener.py (hunk 4 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: Actually it can happen if the collection is deleted and the plugin wasn't activated before the last change.

File: kinto/plugins/quotas/test_plugin.py [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+        data = self.storage.get("quota", self.collection_uri,
+                                "collection_info")
+        self.assertStatsEqual(data, {
+            "record_count": 1,
+            "storage_size": storage_size
+        })
+
+
+class QuotaMaxBytesExceededCollectionSettingsListenerTest(
+        FormattedErrorMixin, QuotaCollectionMixin, QuotaWebTest):
```
Comment (glasserc) [inline kinto/plugins/quotas/test_plugin.py]: This is a clever way to define tests.

File: kinto/plugins/quotas/test_plugin.py [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+
+    def get_app_settings(self, extra=None):
+        settings = super(QuotaMaxBytesPerItemExceededBucketListenerTest,
+                         self).get_app_settings(extra)
+        settings['quotas.bucket_test_max_bytes_per_item'] = '52'
+        return settings
+
+
+class QuotaCollectionMixin(object):
+    def test_507_is_raised_if_quota_exceeded_on_record_creation(self):
```
Comment (glasserc) [inline kinto/plugins/quotas/test_plugin.py]: I guess there's no need to check for 507s on updates?

Comment (Natim) [inline kinto/plugins/quotas/test_plugin.py]: Yes we should add a test for it.

File: kinto/plugins/quotas/test_plugin.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+        self.create_record()
+        body = {'data': {'foo': 42}}
+        resp = self.app.post_json('%s/records' % self.collection_uri,
+                                  body, headers=self.headers, status=507)
+
+        self.assertFormattedError(
+            resp, 507, ERRORS.FORBIDDEN, "Insufficient Storage",
+            "There was not enough space to save the resource")
+
+        # Check that the storage was not updated.
```
Comment (glasserc) [inline kinto/plugins/quotas/test_plugin.py]: Should we verify that the record didn't get created too?

File: kinto/plugins/quotas/listener.py (hunk 1 above)
Comment (glasserc) [inline kinto/plugins/quotas/listener.py]: I think a race condition shouldn't be possible, since the ResourceChanged event happens within a transaction.

File: kinto/plugins/quotas/listener.py (hunk 6 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: I added tests and it seems that because the quota is removed before there it doesn't trigger the quota on delete already.

File: kinto/plugins/quotas/test_plugin.py (hunk 9 above)
Comment (Natim) [inline kinto/plugins/quotas/test_plugin.py]: Tests added.

File: kinto/plugins/quotas/test_plugin.py (hunk 10 above)
Comment (Natim) [inline kinto/plugins/quotas/test_plugin.py]: Yes I am adding a test there.

**Comment by Natim** [comment]
I realized that transaction would only work with the Postgresql backend.

**Comment by Natim** [comment]
Looking at it again, it seems that we need to add additional work:
- [ ] Add documentation
- [ ] Add a test on batch with multiple collection with records delete for instance
- [ ] Add tests for the default bucket

File: docs/api/1.x/quotas.rst [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+Clients can check for the ``quotas`` capability in the
+:ref:`root URL endpoint <api-utilities>`.
+
+.. note::
+
+    In terms of performance, enabling this plugin generates two or
+    three additional queries on backends per request.
+
+* The bucket quota consider the bucket attributes, groups attributes,
```
Comment (glasserc) [inline docs/api/1.x/quotas.rst]: `considers`. I would also reword this, something like "A bucket's quota is a limit on the size of bucket attributes, group attributes, collection attributes, and record attributes."

File: docs/api/1.x/quotas.rst [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+  collections attributes, records attributes.
+* Deleted items size is considered to be null so if you add something
```
Comment (glasserc) [inline docs/api/1.x/quotas.rst]: `Deleted items'`, with the apostrophe to indicate possession (the size of the deleted items). I'd probably reword this as "Deleted items are considered to have zero size" or "size zero".

File: docs/api/1.x/quotas.rst [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+  and remove it, it will looks like it was never created for the
```
Comment (glasserc) [inline docs/api/1.x/quotas.rst]: look like

File: docs/api/1.x/quotas.rst [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+  quota even if its tombstone is still there.
+* The quota plugins only works with the transactional storage backends
```
Comment (glasserc) [inline docs/api/1.x/quotas.rst]: plugin

File: docs/api/1.x/quotas.rst [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+You can configure three types of quotas:
+
+* **QUOTA_BYTES**: The maximum total amount (in bytes) of data that
+  can be stored in sync storage, as measured by the Canonical JSON
+  stringification of every value plus every key's length.
+* **QUOTA_BYTES_PER_ITEM**: The maximum size (in bytes) of each
+  individual item in sync storage, as measured by the Canonical JSON
+  stringification of its value plus its key length.
+* **MAX_ITEMS**: The maximum number of items that can be stored in
+  sync storage.
```
Comment (glasserc) [inline docs/api/1.x/quotas.rst]: That can be stored in a collection or bucket?

File: docs/api/1.x/quotas.rst [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+  ``kinto.quotas.collection_{bucket_name}_max_items`` i.e
+  ``kinto.quotas.collection_blocklists_max_items``
+* Specifically for a given bucket collection using
+  ``kinto.quotas.collection_{bucket_name}_{collection_name}_max_bytes``,
+  ``kinto.quotas.collection_{bucket_name}_{collection_name}_max_bytes_per_item`` and
+  ``kinto.quotas.collection_{bucket_name}_{collection_name}_max_items`` i.e
+  ``kinto.quotas.collection_blocklists_certificates_max_items``
+
+
+How does it works?
```
Comment (glasserc) [inline docs/api/1.x/quotas.rst]: "work", because "it does work"

**Comment by Natim** [comment]
Final r+ @leplatrem

File: CHANGELOG.rst [hunk 17]
```diff
@@ -26,7 +26,8 @@ This document describes changes between each past release.
 - Added ability to plug custom StatsD backend implementations via a new ``kinto.statsd_backend``
   setting. Useful for Datadog™ integration for example (fixes #626).
 - Added a ``delete-collection`` action to the ``kinto`` command. (#727)
-- Added verbosity options to the ``kinto`` command (#745)
+- Added verbosity options to the ``kinto`` command. (#745)
+- Added a quotas plugin. (#752)
```
Comment (leplatrem) [inline CHANGELOG.rst]: nitpick: added a built-in plugin that allows to define quotas per bucket or collection...

File: docs/api/1.x/quotas.rst [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+  (i.e PostgreSQL)
+
+
+Configuration
+=============
+
+You can configure three types of quotas:
+
+* **QUOTA_BYTES**: The maximum total amount (in bytes) of data that
+  can be stored in sync storage, as measured by the Canonical JSON
```
Comment (leplatrem) [inline docs/api/1.x/quotas.rst]: sync storage? canonical json?

File: docs/api/1.x/quotas.rst [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+  stringification of every value plus every key's length.
+* **QUOTA_BYTES_PER_ITEM**: The maximum size (in bytes) of each
+  individual item in sync storage, as measured by the Canonical JSON
+  stringification of its value plus its key length.
+* **MAX_ITEMS**: The maximum number of items that can be stored in
+  a collection or bucket.
+
+You can configure it at multiple levels.
```
Comment (leplatrem) [inline docs/api/1.x/quotas.rst]: in INI settings file

File: docs/api/1.x/quotas.rst [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+For buckets:
+
+* Globally for each buckets using ``kinto.quotas.bucket_max_bytes``,
```
Comment (leplatrem) [inline docs/api/1.x/quotas.rst]: typo: each bucket

File: docs/api/1.x/quotas.rst [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+  ``kinto.quotas.bucket_max_bytes_per_item`` and
+  ``kinto.quotas.bucket_max_items``
+* Specifically for some buckets using
+  ``kinto.quotas.bucket_{bucket_name}_max_bytes``,
```
Comment (leplatrem) [inline docs/api/1.x/quotas.rst]: nitpick: I think we use *bucket id* elsewhere ?

File: docs/api/1.x/quotas.rst [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+  ``kinto.quotas.bucket_{bucket_name}_max_bytes_per_item`` and
+  ``kinto.quotas.bucket_{bucket_name}_max_items`` i.e
```
Comment (leplatrem) [inline docs/api/1.x/quotas.rst]: nitpick: I think it's `e.g.` instead of `i.e.`

File: kinto/plugins/quotas/__init__.py [hunk 23]
```diff
@@ -0,0 +1,14 @@
+from kinto.core.events import ResourceChanged
+
+from .listener import on_resource_changed
+
+
+def includeme(config):
+    config.add_api_capability('quotas',
+                              description='Quotas Management on Buckets.',
```
Comment (leplatrem) [inline kinto/plugins/quotas/__init__.py]: nitpick: *and collections*

File: kinto/plugins/quotas/listener.py [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+        settings.get('quotas.collection_{}_{}'.format(bucket_id, name),
+                     # Global to all buckets collections
+                     settings.get('quotas.collection_{}'.format(name), None)))
+
+
+def on_resource_changed(event):
+    """
+    Everytime an object is created/changed/deleted, we update the
+    bucket counters. The entries are served as read-only in the
+    :mod:`kinto.plugins.quotas.views` module.
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: This doesn't seem to be relevant anymore

File: kinto/plugins/quotas/listener.py [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+
+    If a new object override the quotas, we reject the request.
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: typo: override**s**: But, isn't *exceed* or *go over* instead of override?

File: kinto/plugins/quotas/listener.py [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+                                                    'max_bytes_per_item')
+    collection_max_bytes = get_collection_settings(settings, bucket_id,
+                                                   collection_id, 'max_bytes')
+    collection_max_items = get_collection_settings(settings, bucket_id,
+                                                   collection_id, 'max_items')
+    collection_max_bytes_per_item = get_collection_settings(
+        settings, bucket_id, collection_id, 'max_bytes_per_item')
+
+    # XXX: Maybe we want to differenciate between
+    #      bucket/collection/records/group payload.
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: I don't understand this comment

File: kinto/plugins/quotas/listener.py [hunk 27]
```diff
@@ -0,0 +1,235 @@
...
+        uri = '/'.join(parts)
+
+        old = impacted.get('old', {})
+        new = impacted.get('new', {})
+
+        targets.append((uri, obj_id, old, new))
+
+    try:
+        bucket_info = copy.deepcopy(
+            storage.get("quota", bucket_uri, 'bucket_info'))
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: Would be better to add explicitly parameter names like above

File: kinto/plugins/quotas/listener.py [hunk 28]
```diff
@@ -0,0 +1,235 @@
...
+        bucket_info = {
+            "collection_count": 0,
+            "record_count": 0,
+            "storage_size": 0,
+        }
+
+    collection_info = {
+        "record_count": 0,
+        "storage_size": 0,
+    }
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: Why not follow the same approach as above for `bucket_info`, put this in except ?

File: kinto/plugins/quotas/listener.py [hunk 29]
```diff
@@ -0,0 +1,235 @@
...
+        try:
+            collection_info = copy.deepcopy(
+                storage.get("quota", collection_uri, 'collection_info'))
+        except RecordNotFoundError:
+            pass
+
+    # Update the bucket quotas values for each impacted record.
+    for (uri, obj_id, old, new) in targets:
+        if resource_name == "collection":
+            collection_uri = uri
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: Why this ? (please add comment, `collection_uri` is defined far away)

File: kinto/plugins/quotas/listener.py [hunk 30]
```diff
@@ -0,0 +1,235 @@
...
+        old_size = record_size(old)
+        new_size = record_size(new)
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: Maybe we should not compute these if no `*bytes*` setting is defined

File: kinto/plugins/quotas/listener.py [hunk 31]
```diff
@@ -0,0 +1,235 @@
...
+
+    try:
+        bucket_info = copy.deepcopy(
+            storage.get("quota", bucket_uri, 'bucket_info'))
+    except RecordNotFoundError:
+        bucket_info = {
+            "collection_count": 0,
+            "record_count": 0,
+            "storage_size": 0,
+        }
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: What happens if the plugin is enabled once the bucket already has some data?

Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: We may list that limitation in the docs

File: kinto/plugins/quotas/listener.py (hunk 28 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: Because in case of a bucket you don't have any collection_info but you are using the variable below.

File: kinto/plugins/quotas/listener.py (hunk 29 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: You are right that this might not be sufficient and we should probably get the collection_info here too.

This is to fix the batch on multiple collections for instance.

File: kinto/plugins/quotas/listener.py [hunk 32]
```diff
@@ -0,0 +1,235 @@
...
+                collection_info['storage_size'] -= old_size
+                collection_info['storage_size'] += new_size
+        elif action == 'delete':
+            bucket_info['storage_size'] -= old_size
+            if resource_name == 'collection':
+                bucket_info['collection_count'] -= 1
+                # When we delete the collection all the records in it
+                # are deleted without notification.
+                collection_records, _ = storage.get_all('record',
+                                                        collection_uri)
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: * Add explicit argument names (`collection_id`, `parent_id`)
* How can you get the records if they were deleted then ?

File: kinto/plugins/quotas/listener.py (hunk 30 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: This is optimization yes.

File: docs/api/1.x/quotas.rst [hunk 33]
```diff
@@ -0,0 +1,93 @@
...
+You can configure three types of quotas:
+
+* **QUOTA_BYTES**: The maximum total amount (in bytes) of data that
+  can be stored in sync storage, as measured by the Canonical JSON
+  stringification of every value plus every key's length.
+* **QUOTA_BYTES_PER_ITEM**: The maximum size (in bytes) of each
+  individual item in sync storage, as measured by the Canonical JSON
+  stringification of its value plus its key length.
+* **MAX_ITEMS**: The maximum number of items that can be stored in
+  a collection or bucket.
```
Comment (leplatrem) [inline docs/api/1.x/quotas.rst]: Those upercase constants look weird (taken from storage.sync isn't it ?) Why don't they all start with QUOTA ? Why using them after all ?

File: kinto/plugins/quotas/listener.py (hunk 31 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: Everything that was in the bucket before the quota is "free"

File: kinto/plugins/quotas/listener.py [hunk 34]
```diff
@@ -0,0 +1,235 @@
...
+    for (uri, obj_id, old, new) in targets:
+        if resource_name == "collection":
+            collection_uri = uri
+        old_size = record_size(old)
+        new_size = record_size(new)
+
+        if max_bytes_per_item is not None and action != "delete":
+            if new_size > max_bytes_per_item:
+                message = ("MAX_BYTES_PER_ITEM size exceeded: %d Bytes. "
+                           "MAX_BYTES_PER_ITEM is %d Bytes" % (
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: I don't think we should keep the use of those upercase constants here. Plus, it's QUOTA_BYTES_PER_ITEM

File: kinto/plugins/quotas/listener.py [hunk 35]
```diff
@@ -0,0 +1,235 @@
...
+                             message=message)
+
+    if bucket_max_items is not None:
+        if bucket_info['record_count'] > bucket_max_items:
+            message = ("Bucket max items exceeded: %d records. "
+                       "MAX_ITEMS is %d records" % (
+                           bucket_info['record_count'], bucket_max_items))
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=message)
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: nitpick: use a small helper `raise_insufficient`

File: kinto/plugins/quotas/listener.py [hunk 36]
```diff
@@ -0,0 +1,235 @@
...
+                           collection_info['record_count'],
+                           collection_max_items))
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=message)
+
+    storage.update(parent_id=bucket_uri,
+                   collection_id='quota',
+                   object_id='bucket_info',
+                   record=bucket_info)
```
Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: use constants for `quota` and `bucket_info` values?

File: kinto/plugins/quotas/test_plugin.py [hunk 37]
```diff
@@ -0,0 +1,889 @@
...
+        self.storage.delete(parent_id='/buckets/test',
+                            collection_id='quota',
+                            object_id='bucket_info')
+        transaction.commit()
+        self.app.delete(self.bucket_uri, headers=self.headers)
+
+    #
+    # Collection
+    #
+    def test_stats_are_not_accessible_if_collection_does_not_exists(self):
```
Comment (leplatrem) [inline kinto/plugins/quotas/test_plugin.py]: typo: does not exist

File: kinto/plugins/quotas/listener.py (hunk 32 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: > How can you get the records if they were deleted then ?

This is because they were not deleted in this yet.

File: docs/api/1.x/quotas.rst (hunk 33 above)
Comment (Natim) [inline docs/api/1.x/quotas.rst]: Maybe @tarekziade have got a point about this.

File: kinto/plugins/quotas/test_plugin.py [hunk 38]
```diff
@@ -0,0 +1,889 @@
...
+        self.assertStatsEqual(data, {
+            "collection_count": 1,
+            "record_count": 3,
+            "storage_size": 232
+        })
+
+    def test_bulk_delete(self):
+        self.create_bucket()
+        self.create_collection()
+        self.create_record()
```
Comment (leplatrem) [inline kinto/plugins/quotas/test_plugin.py]: Why do you create a collection and a record here ?

Comment (Natim) [inline kinto/plugins/quotas/test_plugin.py]: I did that in order to validate that DELETE works properly.

File: kinto/plugins/quotas/listener.py (hunk 35 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: Actually I find it easier to read like that because it removes a level of indirection and make it explicit that it will raise. I don't think creating a level of indirection with only another function call adds much value here.

File: kinto/plugins/quotas/listener.py (hunk 34 above)
Comment (Natim) [inline kinto/plugins/quotas/listener.py]: I find it very explicit, how would you like it to be?

Comment (leplatrem) [inline kinto/plugins/quotas/listener.py]: I guess that those upercase strings refer to the ones in the docs ? `QUOTA_BYTES_PER_ITEM` or `QUOTA_BYTES` ? But you used `MAX_BYTES_PER_ITEM` and `MAX_BYTES`. Why not just something like `Maximum bytes size per item exceeded (%d > %d Bytes)` ? (*not crucial, but while we're at it..*)

File: docs/api/1.x/quotas.rst [hunk 39]
```diff
@@ -0,0 +1,93 @@
...
+    three additional queries on backends per request.
+
+* A bucket's quota is a limit on the size of bucket attributes, group
+  attributes, collection attributes, and record attributes.
+* Deleted items are considered to have a size zero so if you add something
+  and remove it, it will look like it was never created for the
+  quota even if its tombstone is still there.
+* The quota plugin only works with the transactional storage backends
+  (e.g. PostgreSQL)
+
```
Comment (leplatrem) [inline docs/api/1.x/quotas.rst]: Note: add mention of handling of existing data when plugin is enabled

File: docs/api/1.x/quotas.rst [hunk 40]
```diff
@@ -0,0 +1,93 @@
...
+You can configure three types of quotas:
+
+* **QUOTA_BYTES**: The maximum total amount (in bytes) of data that
+  can be stored in a bucket or collection, as measured by the JSON
+  stringification of every value plus every key's length.
+* **QUOTA_BYTES_PER_ITEM**: The maximum size (in bytes) of each
+  individual item in the bucket or collection, as measured by the JSON
+  stringification of its value plus its key length.
+* **MAX_ITEMS**: The maximum number of items that can be stored in
+  a collection or bucket.
```
Comment (leplatrem) [inline docs/api/1.x/quotas.rst]: Those uppercase strings come from https://developer.chrome.com/apps/storage. Not sure if they bring anything compared to simple bold string (`Maximum total bytes`, `Maximum bytes per object`, `Maximum number of objects`...).

I realize that we use *objects* in Kinto and not *items*

File: kinto/plugins/quotas/listener.py (hunk 6 above)
Comment (glasserc) [inline kinto/plugins/quotas/listener.py]: Try adding a test where the quota is reduced after a bunch of objects have been stored, and then doing a delete.
